assign from shift

The following program `sample.go` triggers a panic:

```go
package main

type H struct {
    bits uint
}

func main() {
    h := &H{8}
    var x uint = (1 << h.bits) >> 6

    println(x)
}
```

Expected result:
```console
$ go run ./sample.go
4
```

Got:
```console
$ yaegi ./sample.go
run: ./sample.go:9:15: CFG post-order panic: reflect: call of reflect.Value.Interface on zero Value
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc0002e0600, 0x40)
        /usr/lib/go/src/runtime/debug/stack.go:24 +0x9f
github.com/traefik/yaegi/interp.(*Interpreter).eval.func1(0xc0001e3c70)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:480 +0xc5
panic(0xeab440, 0xc000209020)
        /usr/lib/go/src/runtime/panic.go:969 +0x175
github.com/traefik/yaegi/interp.(*Interpreter).cfg.func2.1(0xc000463b00)
        /home/marc/go/src/github.com/traefik/yaegi/interp/cfg.go:424 +0xb3
panic(0xe01b20, 0xc000208fe0)
        /usr/lib/go/src/runtime/panic.go:969 +0x175
reflect.valueInterface(0x0, 0x0, 0x0, 0x10a1601, 0xdccd60, 0x1)
        /usr/lib/go/src/reflect/value.go:1020 +0x1ad
reflect.Value.Interface(...)
        /usr/lib/go/src/reflect/value.go:1015
github.com/traefik/yaegi/interp.typecheck.shift(0xc000463b00, 0x7fafa5972280, 0x0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/typecheck.go:152 +0x552
github.com/traefik/yaegi/interp.typecheck.binaryExpr(0xc000463b00, 0xc00030a960, 0xef901c)
        /home/marc/go/src/github.com/traefik/yaegi/interp/typecheck.go:223 +0x547
github.com/traefik/yaegi/interp.(*Interpreter).cfg.func2(0xc000463b00)
        /home/marc/go/src/github.com/traefik/yaegi/interp/cfg.go:633 +0x8ffd
github.com/traefik/yaegi/interp.(*node).Walk(0xc000463b00, 0xc0001e3af0, 0xc0001e3ac0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:220 +0xa5
github.com/traefik/yaegi/interp.(*node).Walk(0xc000463800, 0xc0001e3af0, 0xc0001e3ac0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:217 +0x66
github.com/traefik/yaegi/interp.(*node).Walk(0xc000463700, 0xc0001e3af0, 0xc0001e3ac0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:217 +0x66
github.com/traefik/yaegi/interp.(*node).Walk(0xc000463600, 0xc0001e3af0, 0xc0001e3ac0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:217 +0x66
github.com/traefik/yaegi/interp.(*node).Walk(0xc000462f00, 0xc0001e3af0, 0xc0001e3ac0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:217 +0x66
github.com/traefik/yaegi/interp.(*node).Walk(0xc000462a00, 0xc0001e3af0, 0xc0001e3ac0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:217 +0x66
github.com/traefik/yaegi/interp.(*node).Walk(0xc000462000, 0xc0001e3af0, 0xc0001e3ac0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:217 +0x66
github.com/traefik/yaegi/interp.(*Interpreter).cfg(0xc00035e000, 0xc000462000, 0xc0002178e8, 0x4, 0xc0002178e8, 0x4, 0x0, 0x0, 0xc000462000)
        /home/marc/go/src/github.com/traefik/yaegi/interp/cfg.go:63 +0x20f
github.com/traefik/yaegi/interp.(*Interpreter).eval(0xc00035e000, 0xc0002babd0, 0x87, 0x7ffc443d9498, 0x19, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:507 +0x2f3
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0xc00035e000, 0x7ffc443d9498, 0x19, 0xc0002bab00, 0x87, 0x0, 0x0, 0xc000212c30)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:394 +0xfa
main.runFile(0xc00035e000, 0x7ffc443d9498, 0x19, 0xc000200180, 0x1)
        /home/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:123 +0xae
main.run(0xc000128030, 0x1, 0x1, 0x13, 0x13)
        /home/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:89 +0x7e5
main.main()
        /home/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:144 +0x415
```

--- a/_test/assign16.go
+++ b/_test/assign16.go
@@ -1,0 +1,15 @@
+package main
+
+type H struct {
+	bits uint
+}
+
+func main() {
+	h := &H{8}
+	var x uint = (1 << h.bits) >> 6
+
+	println(x)
+}
+
+// Output:
+// 4

--- a/_test/select1.go
+++ b/_test/select1.go
@@ -1,18 +1,20 @@
 package main
 
-import "time"
-import "fmt"
+import (
+	"fmt"
+	"time"
+)
 
 func main() {
 	c1 := make(chan string)
 	c2 := make(chan string)
 
 	go func() {
-		time.Sleep(1e9)
+		time.Sleep(1e7)
 		c1 <- "one"
 	}()
 	go func() {
-		time.Sleep(2e9)
+		time.Sleep(2e7)
 		c2 <- "two"
 	}()
 

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -677,7 +677,6 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 				n.gen = nop
 				n.findex = -1
 			case n.anc.kind == assignStmt && n.anc.action == aAssign:
-				//case (n.anc.kind == assignStmt || n.anc.kind == defineStmt) && n.anc.action == aAssign:
 				// To avoid a copy in frame, if the result is to be assigned, store it directly
 				// at the frame location of destination.
 				dest := n.anc.child[childPos(n)-n.anc.nright]
@@ -2459,7 +2458,7 @@ func isValueUntyped(v reflect.Value) bool {
 	return t.String() == t.Kind().String()
 }
 
-// isArithmeticAction returns true if the node action is an arithmetic operator
+// isArithmeticAction returns true if the node action is an arithmetic operator.
 func isArithmeticAction(n *node) bool {
 	switch n.action {
 	case aAdd, aAnd, aAndNot, aBitNot, aMul, aQuo, aRem, aShl, aShr, aSub, aXor:

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -551,6 +551,14 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 					n.gen = nop
 					src.findex = dest.findex
 					src.level = level
+				case n.action == aAssign && len(n.child) < 4 && !src.rval.IsValid() && isArithmeticAction(src):
+					// Optimize single assignments from some arithmetic operations.
+					// Skip the assign operation entirely, the source frame index is set
+					// to destination index, avoiding extra memory alloc and duplication.
+					src.typ = dest.typ
+					src.findex = dest.findex
+					src.level = level
+					n.gen = nop
 				case src.kind == basicLit && !src.rval.IsValid():
 					// Assign to nil.
 					src.rval = reflect.New(dest.typ.TypeOf()).Elem()
@@ -669,6 +677,7 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 				n.gen = nop
 				n.findex = -1
 			case n.anc.kind == assignStmt && n.anc.action == aAssign:
+				//case (n.anc.kind == assignStmt || n.anc.kind == defineStmt) && n.anc.action == aAssign:
 				// To avoid a copy in frame, if the result is to be assigned, store it directly
 				// at the frame location of destination.
 				dest := n.anc.child[childPos(n)-n.anc.nright]
@@ -2448,4 +2457,14 @@ func isValueUntyped(v reflect.Value) bool {
 		return true
 	}
 	return t.String() == t.Kind().String()
+}
+
+// isArithmeticAction returns true if the node action is an arithmetic operator
+func isArithmeticAction(n *node) bool {
+	switch n.action {
+	case aAdd, aAnd, aAndNot, aBitNot, aMul, aQuo, aRem, aShl, aShr, aSub, aXor:
+		return true
+	default:
+		return false
+	}
 }

--- a/interp/interp_eval_test.go
+++ b/interp/interp_eval_test.go
@@ -986,6 +986,9 @@ func TestConcurrentEvals(t *testing.T) {
 // called by EvalWithContext is sequential. And that there is no data race for the
 // interp package global vars or the interpreter fields in this case.
 func TestConcurrentEvals2(t *testing.T) {
+	if testing.Short() {
+		return
+	}
 	pin, pout := io.Pipe()
 	defer func() {
 		_ = pin.Close()
@@ -1045,6 +1048,9 @@ func TestConcurrentEvals2(t *testing.T) {
 // - when calling Interpreter.Use, the symbols given as argument should be
 // copied when being inserted into interp.binPkg, and not directly used as-is.
 func TestConcurrentEvals3(t *testing.T) {
+	if testing.Short() {
+		return
+	}
 	allDone := make(chan bool)
 	runREPL := func() {
 		done := make(chan error)
@@ -1123,6 +1129,9 @@ func TestConcurrentComposite2(t *testing.T) {
 }
 
 func testConcurrentComposite(t *testing.T, filePath string) {
+	if testing.Short() {
+		return
+	}
 	pin, pout := io.Pipe()
 	i := interp.New(interp.Options{Stdout: pout})
 	i.Use(stdlib.Symbols)
@@ -1160,6 +1169,9 @@ func testConcurrentComposite(t *testing.T, filePath string) {
 }
 
 func TestEvalScanner(t *testing.T) {
+	if testing.Short() {
+		return
+	}
 	type testCase struct {
 		desc      string
 		src       []string
@@ -1333,6 +1345,9 @@ func applyCIMultiplier(timeout time.Duration) time.Duration {
 }
 
 func TestREPLDivision(t *testing.T) {
+	if testing.Short() {
+		return
+	}
 	_ = os.Setenv("YAEGI_PROMPT", "1")
 	defer func() {
 		_ = os.Setenv("YAEGI_PROMPT", "0")

--- a/interp/typecheck.go
+++ b/interp/typecheck.go
@@ -486,7 +486,7 @@ func (check typecheck) sliceExpr(n *node) error {
 	case reflect.Array:
 		valid = true
 		l = t.Len()
-		if c.kind != selectorExpr && c.kind != indexExpr && (c.sym == nil || c.sym.kind != varSym) {
+		if c.kind != indexExpr && c.kind != selectorExpr && (c.sym == nil || c.sym.kind != varSym) {
 			return c.cfgErrorf("cannot slice type %s", c.typ.id())
 		}
 	case reflect.Slice:

--- a/interp/typecheck.go
+++ b/interp/typecheck.go
@@ -151,7 +151,7 @@ func (check typecheck) shift(n *node) error {
 	t0, t1 := c0.typ.TypeOf(), c1.typ.TypeOf()
 
 	var v0 constant.Value
-	if c0.typ.untyped {
+	if c0.typ.untyped && c0.rval.IsValid() {
 		v0 = constant.ToInt(c0.rval.Interface().(constant.Value))
 		c0.rval = reflect.ValueOf(v0)
 	}
@@ -486,7 +486,7 @@ func (check typecheck) sliceExpr(n *node) error {
 	case reflect.Array:
 		valid = true
 		l = t.Len()
-		if c.kind != selectorExpr && (c.sym == nil || c.sym.kind != varSym) {
+		if c.kind != selectorExpr && c.kind != indexExpr && (c.sym == nil || c.sym.kind != varSym) {
 			return c.cfgErrorf("cannot slice type %s", c.typ.id())
 		}
 	case reflect.Slice:

